### ISSUE TITLE
Support multi-level density for all components

### ISSUE THREAD

**Issue by jelbourn** [angular/components#4597]
This issue tracks the ability to specify a density level for components. This would include:
* Uniform density API for all components
* Predetermined density levels, configurable via an approach similar to theming
* Configurable default density
* Consistent appearance for components with the same density level

This is planned work for after beta ends.

**Comment by code-tree** [comment]
Just a note for icons that Material guidelines specifies:

> Clearance area (non-dense)
> Icon: 24dp
> Touch target: 48dp
> 
> Clearance area for dense icons
> Icon: 20dp
> Touch target: 40dp

Angular Material seems to currently use a hybrid for icon buttons of 24px icon size (non-dense) and 40px touch target (dense).

On top of that, the official Material icons seem to only come in 24px (non-dense) size anyway? Is it worth switching consistently to just non-dense now, even before a density API is created?

**Comment by jaa1982** [comment]
Any update on ETA for this ?  As a developer of enterprise apps, 95% of our usage is on the desktop, and spacing / density has been our users primary concern since first adopting the material spec with angular material v1.0. 

“Dense mode” features had been previously scheduled for angular-material v1.2 but later fell out of scope, as momentum shifted to this repo (  https://github.com/angular/material/issues/9201 ) .  Now as we make the transition to material v2, it would be nice to have confirmation that “dense mode” is still a priority item and slated for work.

**Comment by jaa1982** [comment]
Hi @jelbourn  , can you please provide some update on plans for high-density features? I notice the post-beta label was removed on this ticket around the same time as v5.0 was being released. Was removing the label a housekeeping change, or does that imply something more significant (i.e. scope change/delay) ? As an enterprise developer targeting 95% usage on the desktop, I'd very much appreciate some insight as I've been closely monitoring this ticket since it was first opened

**Comment by ganeshkantu** [comment]
+1

**Comment by dhniels** [comment]
@jaa1982 did you ever get any updates on this? I'd like to follow this issue as well. having a dense table option would be nice. seems like the default rows are slightly on the airy side for some of our larger data-filled tables.

**Comment by jaa1982** [comment]
Hi @dhniels , no, no updates from the team on this as of yet. Our own team has considered overriding some of the default styles ourselves, but we're reluctant because of potential for breaking changes down the road and/or just not getting it right.

I'm wondering if @jelbourn or others on the material team might have a recommendation for what we might do in the meantime. For example, for denser forms and tables, should we be overriding default typography sizes? 

The spec's not clear on typography sizes for dense layouts , but you can refer to the current defaults here:

https://github.com/angular/material2/blob/master/src/lib/core/typography/_typography.scss#L30

We're thinking we might scale it down a bit like this:

  $body-2:        mat-typography-level(12px, 20px, 500),
  $body-1:        mat-typography-level(12px, 18px, 400),
  $caption:       mat-typography-level(10px, 18px, 400),
  $button:        mat-typography-level(12px, 12px, 500),
  // Line-height must be unit-less fraction of the font-size.
  $input:          mat-typography-level(12px, 1.125, 400)

Of course any and all advice would be appreciated

**Comment by palpatine1991** [comment]
We would appreciate too some estimated timeline. We are also tight on the schedules and thinking about creating some implementation of dense mode by ourselves (but also concerning about breaking changes)

**Comment by adoris** [comment]
Many enterprise developers extremely waitig for high-density features in material components! 
Can you please provide some update on plans for high-density features?
Thanks!

**Comment by jelbourn** [comment]
Still something we want to do, but hasn't been as high of a priority as other features.

**Comment by adoris** [comment]
there are some roadmap for "not high priority" features like this?

**Comment by dhniels** [comment]
this can be easily accomplished with custom css. thanks @jelbourn and team for considering the feature - great work on the project so far. Looking forward to continued progress in the repo.

**Comment by jelbourn** [comment]
We typically decide what to work on on a quarterly basis

**Comment by krag-123** [comment]
Is this still “not high priority”?

**Comment by amitoj** [comment]
If this is not considered very soon, or provided a work around with example, I’ll have no option but to move away from material design to either Bootstrap or a 3rd party UI components provider. 
Sad this simple request is lingering on for this long.

**Comment by jrovny** [comment]
I'd also like to voice my desire for dense properties for Material controls. At my company we have large, complex forms so dense, tight controls with minimal padding/margins is very important to us.

**Comment by amitoj** [comment]
Managed to achieve dense mat-table with folloing in styles.scss 
`.mat-table > .mat-header-row, .mat-table > .mat-row {
  min-height: 34px;
}`

**Comment by amitoj** [comment]
Managed dense mat-table with folloing in styles.scss
`.mat-table > .mat-header-row, .mat-table > .mat-row { min-height: 34px; }`

**Comment by zakdances** [comment]
Simply overriding the styles of a subcomponents like `mat-row` isn't a great workaround because it leaves other subcomponents with disproportionate dimensions/padding/font sizes/etc. For example, you can change the table rows to `34px` but that leaves the header row at `56px` which now looks comically large in comparison. So now you have to change the header row. But to what? Should you try to maintain the same proportion? Which would mean the header now needs to be a `38.8571428572px`. So now does the table's bottom pagination row need to be changed too? The button spacing?

It's even worse with components whose many children depend on negative margins and custom padding (like `mat-card`).

It'd be really great to have at least 3 official density options (`normal`, `lowDense`, `highDense`) on all top-level components. (especially `mat-card`! There's so much styling there that needs to be overridden)

**Comment by thw0rted** [comment]
I wouldn't expect support for anything that isn't in the spec.  If you want a particular component to support high/low density, you might need to take it up with Google.

**Comment by zakdances** [comment]
@thw0rted angular/material2 is an official Google repo, but I think you meant take it up with the Material Design team specifically. True, a low density mode isn't in spec, I just think it'd be very useful. High density should be very much in the pipeline though.

**Comment by Nodarii** [comment]
@jelbourn , 

Is it possible to increase priority for this issue? 
I see that you have initiated this discussion before it appeared in official material design spec (probably that's why it had low priority). 
Now it is part of material 2.0 specification and I guess you should increase priority for issues that aligns angular-material to the material specification.

**Comment by jelbourn** [comment]
The current status is that some component define ad-hoc densities in the spec. We're like to get the designers to unify on a consistent scheme for density that spans all of the components. It's something we're talking about, but still no ETA on implementation.

**Comment by igorkemack** [comment]
Is there any updates?

**Comment by jelbourn** [comment]
We hope to have something here in early 2020.

**Comment by jneuhaus20** [comment]
The [Material Design spec](https://material.io/design/layout/applying-density.html#) says (high) density should apply somewhat narrowly, e.g. form fields, list rows, buttons change, but date pickers, dialogs, and anything that could be a touch target shouldn't.  Those are nice boundaries with design philosophy support for saying "you're on your own" outside of them.

Honestly, I think a lot of us would be happy if we could at least apply some mixins (a la `angular-material-theme`.)  Material's use of variables is really well architected, and from looking at `_form-field-outline-theme.scss` and `form-field-outline.scss`, most (but not all) of what we need is already in there.

Even just guidance on how (if) we can close the gap on our own while we wait for proper all-encompassing mixins would be a huge boon.

**Comment by the-ult** [comment]
@jelbourn what is the status of the dense variants? (Since your comment on the 13th of june? )

Is it gonna be part of the MDC specs (new material-experimental), or is it gonna be implemented before that in the current Angular Material?

**Comment by jelbourn** [comment]
It will be part of the new MDC-based versions of the components

**Comment by edusoccer1121** [comment]
> It will be part of the new MDC-based versions of the components

Could you explain what this means? Should I wait for this library to support this or will I need to switch to another components library? Should those "new" components will be merge to this one or what? I am really confused about the future of this feature.

**Comment by Robinyo** [comment]
![Screen Shot 2020-01-01 at 09 33 59](https://user-images.githubusercontent.com/12203540/71635587-0503f880-2c7a-11ea-8903-5374035fe091.png)

Ref:
- https://material-density.glitch.me/
- https://github.com/material-components/material-components-web/tree/master/packages/mdc-density
- https://material.io/design/layout/applying-density.html#

**Comment by HunderlineK** [comment]
I created a ticket for this about three years ago:
https://github.com/angular/components/issues/4553

And still not a high priority... the joke is that the toolbar at material.angular.io is dense itself.

Good job listening to the community lol

**Comment by talamaska** [comment]
> It will be part of the new MDC-based versions of the components

I see a bunch of components in the experimental folder, still missing a few, but still can't see anything related to the dense mode.

**Comment by JohnnyQQQQ** [comment]
The angular components team is actively working on this right now https://github.com/angular/components/tree/density-api

**Comment by Robinyo** [comment]
Some additional links:
- [Using Material Density on the Web](https://medium.com/google-design/using-material-density-on-the-web-59d85f1918f0)
- [Material Components Web (MDC Web) - Density](https://github.com/material-components/material-components-web/tree/master/packages/mdc-density)
- [Integrating MDC Web into Frameworks](https://material.io/develop/web/docs/framework-integration/)

**Comment by thw0rted** [comment]
I don't fully grok the relationship between `material-components` and this project.  Aren't both supported by Google? And the former has a [third-party Angular adapter](https://github.com/trimox/angular-mdc-web)?

**Comment by jelbourn** [comment]
The MDC primitives are created by the Material Design team, whereas Angular Material is made by the Angular team. MDC started much after Angular Material was created, so now we're working on integrating that work into the Angular version, but rolling this out without just completely breaking everyone is a lot of work.

**Comment by thw0rted** [comment]
Thanks Jeremy, that helps.  Is the end goal for one of the projects to get rolled into the other?  Does it make sense for end-users to start looking at migrating?  It's probably not worth the hassle just for one feature like first-class density support.

**Comment by jelbourn** [comment]
MDC Web is really just a set of primitives- they don't come with a "component" primitive. Angular Material composes those primitives into a component that participates in Angular's framework idioms. Density is just the flashiest feature- the main benefit is that the Material Design team will take on ownership of making sure everything is on-spec and up-to-date with the spec (which is difficult for us because there's no actual changelog or update channel for the spec).

**Comment by coder925** [comment]
If [Angular Material](https://github.com/angular/components) will compose/wrap [MDC Web](https://github.com/material-components/material-components-web) primitives, will I be able to directly reference the styles from MDC Web (e.g. `<div class="mdc-data-table">`) ?

I’m using Angular Material but need to build one custom data table from scratch with specific behaviour (cannot use [Table component](https://material.angular.io/components/table/overview)). Still, I want to reuse the table styles. Can this be done solely with Angular Material or do I need to bring in MDC Web as well?

**Comment by jelbourn** [comment]
The CSS will be present, but I can't say now how much of it will be brought in.

**Comment by awallat** [comment]
I saw in the changelog for version 10.0.0 the following features for material-experimental:

> mdc-form-field: support density scaling
...

I thought this would close this issue.

Is there any documentation how to use this density scaling? I want to easily change the size of input fields, buttons, and so on.

**Comment by benelliott** [comment]
@awallat That commit is specific to the new implementation of components that are based on MDC. These new components are still under active development so they aren't referred to in docs yet. But I expect that the density features will be documented when the MDC components eventually reach production.

**Comment by brmc** [comment]
@benelliott, do you have a rough estimation of when it might be released? I'm in the process of vetting component libraries and the low density is one of the major strikes against material. it's very pleasant to use, but the low density is a no-go for our use case

**Comment by benelliott** [comment]
@brmc I don't work on the team so can only point you to the [README](https://github.com/angular/components/blob/master/README.md) which still lists it as a high-priority item for Q3.

**Comment by jobinjamez** [comment]
@jelbourn How realistic are plans to release this by Q2? We are building our design system using Angular material as the base. We are theming it quite much and I would like to test if our theming breaks density in some way. Is there some documentation on how to use it from the experimental branch?

**Comment by jelbourn** [comment]
I don't think we've ever set a concrete ETA. The mdc components that support density are usable from the `material-experimental` package, though these are not yet considered stable. We're working on rolling the new versions out internally inside Google, and using the feedback from that to fix bugs and inconsistencies before we get to that stable milestone. We don't know 100% what kinds of issues we'll run into as we do the rollout, so we can't say when exactly the process will be done.

**Comment by jobinjamez** [comment]
Thank you @jelbourn, Did I understand correctly that you have density support implemented both for Angular components (e.g Tree) and the Angular MDC components? Which version do you plan to work on further? Is the density usage for MDC components the same as in the angular components? i.e passing density value (-1,-2 & -3) to the theme.

**Comment by jelbourn** [comment]
The density system will only apply to the new MDC-based components plus a few other components that don't have MDC equivalents (like expansion panel). This is because we based our density APIs on top of the density system built by the MDC team rather than making something from scratch. This is part of the reason why we're moving over to use MDC in the first place- it lets us take advantage of a larger pool of work at the cost of the one-time switchover.

**Comment by JanneHarju** [comment]
So is angular material components just ditched? So never density support or any kind of better material design support? Or is it just  in low priority even it has P2 label on in it (Important issue that should be resolved as soon as possible). This has been waited for too long. Almost very first steps with Angular Material when I started with it about 3 years ago I started suspect was decision correct to choose Material over Bootstrap. Or will threre be somekind of Angular(11) MDC library coming?

**Comment by albyrock87** [comment]
@jelbourn when MCD-based components will be merged into master, will current components/scss be still available?
We made a lot of effort on customizing current components (via CSS overrides).
Thanks

**Comment by jneuhaus20** [comment]
@JanneHarju 
> new MDC-_based_ components [emphasis added]

> ...we're moving over to use MDC

> ...at the cost of the one-time switchover

They are replacing the underlying implementation, not abandoning Material Components.

**Comment by jneuhaus20** [comment]
@albyrock87 They're already in master, just in a different package.

We made a number of customizations as well, including relying on some internal variable values we re-exposed (for consistency) and altering the density somewhat.  I imagine there will be little disruption when completing the switch, because that would break a lot of theming, but it would still behoove us to test first.

**Comment by cgatian** [comment]
@jelbourn You and your team are doing amazing work. I can imagine how difficult this must be. Thank you! 🙏

**Comment by JanneHarju** [comment]
> @JanneHarju
> 
> > new MDC-_based_ components [emphasis added]
> 
> > ...we're moving over to use MDC
> 
> > ...at the cost of the one-time switchover
> 
> They are replacing the underlying implementation, not abandoning Material Components.

Sorry for my last comment. I missundestood that earlier comment. Hopefully our css customization and theming won't need to change when MDC change occures.

**Comment by Baffour** [comment]
> > @JanneHarju
> > > new MDC-_based_ components [emphasis added]
> > 
> > 
> > > ...we're moving over to use MDC
> > 
> > 
> > > ...at the cost of the one-time switchover
> > 
> > 
> > They are replacing the underlying implementation, not abandoning Material Components.
> 
> Sorry for my last comment. I missundestood that earlier comment. Hopefully our css customization and theming won't need to change when MDC change occures.

@JanneHarju It's worth looking into a visual regression solution imo. Something like [this](https://blog.logrocket.com/automated-visual-regression-testing-with-jest/) or [this](https://storybook.js.org/docs/angular/workflows/visual-testing). That way when it comes to migrating later on you can at least see what breaks without having to eyeball it

**Comment by Baffour** [comment]
Hey @jelbourn I get you can't give an ETA or anything, but just wondering what the visibility of progress is like? I've seen this project https://github.com/angular/components/projects/25 but doesn't seem to cover your internal rollout + none of them have been updated for a while (or is that just an accurate reflection of progress?)

**Comment by jelbourn** [comment]
The status today is that most of the components are done and fully functional except for a handful where the MDC version wasn't ready until recently. The tooltip is landing soon (MDC just finished it themselves), slider is in progress (which will include range slider support), list will be worked on soon, and we're waiting for toggle-button to be stable on the MDC before we can tackle that. 

As we roll the components out inside Google, we're catching bugs and making small API changes where necessary to cut down on impedance mismatch. For example, one thing that came up recently was that `MatFormField` has a generic container for prefix/suffix, but MDC has separate prefix/suffix styles for text and icons. So, we're tweaking the API a bit to align more with MDC. It's hard to know all of the little rough edges like this we'll run into without going through the rollout process, which is what makes giving progress difficult.

**Comment by Baffour** [comment]
Thanks @jelbourn really helpful (and interesting!) to know 😁

**Comment by cluen** [comment]
@jelbourn Can you please give a quick update on this topic? Your last information is very helpful but 2 months old ;-) Are there any changes. Can't wait to use density with angular-material components.

thanks a lot!

**Comment by jelbourn** [comment]
I still don't have an ETA on moving everything to stable. We're continuing to push on rolling our mdc-based components in Google, but there's still a ways to go.

**Comment by niccolofanton** [comment]
@jelbourn how is the project going? Do you have any updates for us?😸  Thank you!

**Comment by jelbourn** [comment]
Nothing to report yet. We're still working away at moving internal Google apps to the MDC-based components (there are over 2k apps, so it's a long road).

**Comment by klemenoslaj** [comment]
> Nothing to report yet. We're still working away at moving internal Google apps to the MDC-based components (there are over 2k apps, so it's a long road).

Just for our understanding. You plan to release the new version only after everything at Google is migrated already?

As well I think we'd be happy already if the MDC-based components follow the deprecations (breaking release could be much more frequent than stable version of course), so that we can start slowly migrating as well, component by component.

I am afraid a lot of projects are hacking/overwritting the styles right now so that might be a bang for them.

**Comment by jelbourn** [comment]
Not _all_ of the applications, just enough that we're satisfied that the components are fully stable and satisfying all use-cases. I know there's a lot of demand for the new versions; we're dedicated to making sure that anything we move to stable is up to our quality standards, which unfortunately takes time.

**Comment by jneuhaus20** [comment]
How are you providing access during the rollout? E.g. is this an internal package so they can use the same selectors/modules after release, or are they importing experimental modules and committing to another round of code updates post-release?

 Given the level of interest, I'm sure a lot of people are in the same boat and would like to follow suit if possible. We added scss to fill this need for example, and can tolerate the risk of using almost-stable components to get ahead on migration and testing. I just don't know the "easy way" to do this that avoids changing all the selectors twice.

**Comment by jelbourn** [comment]
The mdc components live in a separate directory, and we have a migration script that swaps from original component to MDC-based component. Since the APIs are mostly identical, we only really need to update the import paths and the Sass mixin names. (Google doesn't have packages like npm since [everything is one one massive monorepo](https://research.google/pubs/pub45424/))

When we do move these components from experimental to stable, we'll also have a schematic for switching in the same way.

**Comment by niccolofanton** [comment]
Is [this board (MDC-based components)](https://github.com/angular/components/projects/25) up to date? @jelbourn

**Comment by jelbourn** [comment]
For the most part, yeah, but it's also not tracking anything related to our rollout of the MDC-based components inside Google.   Once that rollout is further along we'll be moving the MDC components to stable.

**Comment by Neutrino-Sunset** [comment]
"the joke is that the toolbar at material.angular.io is dense itself"

Yes, it appeared they had to handroll a custom component for their own documentation site instead of using their own library.

I don't see any mention of Toolbar here or in any of the links provided so far. Is Toolbar going to have dense support when this finally comes? That is one the original tickets that was closed in favour of this one, and dense appbar support on desktop has been specifed in Material Design from the outset.

**Comment by MichelReij** [comment]
I am also very curious and anxiously waiting for updates on the density.

**Comment by Microhive** [comment]
I am also eagerly awaiting this feature. Would be great to have an update.

**Comment by davidecampello** [comment]
me too!

**Comment by lebesnec** [comment]
is this feature dead? Is anyone still working on Angular Material?
it's been >4 years since this ticket was open, and >4 months since the last update from a team member.

**Comment by Waterstraal** [comment]
@mgechev said the following in the latest _State of Angular Ecosystem_ talk:

> "We'll be kicking off our rollout plan around MDC Web early next year as well"
https://youtu.be/kZXUFKqWrIY?t=2024

I'm curious to what that exactly means. Could you please provide more details @mgechev?

Thanks!

**Comment by mgechev** [comment]
We're coordinating the effort with the MDC team at the moment. Looks like we'll be unable to release this work as part of v14. Currently we're looking at v15+ or later.

**Comment by stevebor1** [comment]
Coming from Quasar Framework (vuejs), which has full Material support with dense and sparse options,  to Angular / Material, I am astonished that this is not supported out of the box :/

**Comment by mjedwabn** [comment]
I am getting increasingly convinced that web frameworks era is ending. They had their time couple of years ago, with adoption of ~components, redux, reactivex etc. Personally Angular lost its momentum for me, I cannot find a good argument to stay with vs what standards offer nowadays. 

I think that pure html/css + typescript + light templating library like _lit_ is he the way to go in the nearest future.

**Comment by talamaska** [comment]
This is not the place to hate about angular. Lit is great but I would love to see you making a large scale project with lit, cause I have seen a polymer one and it is not pretty.

**Comment by achacofe** [comment]
Good Morning, reading this issue I think that the new mdc theme is what I am looking for for a case of Angular applications in which they are displayed in a resolution of 1920px x 1080px at a scale of 125%, 150% and 200%, since With these scales obviously the components look bigger to me, for the reason that the viewport is reduced:

1. 125% -> 1536px x 850px

2. 150% -> 1280px x 690px

3. 200% -> 960px x 490px

This is also due to the need of the application, in my case I have several applications that with these scales do not want a scroll on the screen, that is, they want everything to be seen on the screen at that scale, as if it were 100%, a design of desktop 1920 x 1080, after several investigations we had to make a patch based on media queries to detect this type of screens and reduce both the font size and padding of the components and others. I do not see this as correct, since it should never be designed for these % of zoom and it is always designed at 100%, but it is a problem that we have spent months looking for the best solution since we are not convinced by the one we have applied.

If anyone else is facing this issue could we open a discussion?

On the other hand, do you think that the new mdc that you comment on in this thread would solve my problem? In that case, how could I implement it for a proof of concept? I need to understand this a little more.

Thank you very much!

**Comment by lixaotec** [comment]
hope it gets solved soon

**Comment by jelbourn** [comment]
It has shipped!

https://material.angular.io/guide/theming#customizing-density

**Comment by stevebor1** [comment]
Does anyone know if density along with all appearance types and inputs (mix and match) display the same height/size in 15? ie fit and display inline correctly?

**Comment by Baffour** [comment]
Amazing, congrats!

**Comment by lebesnec** [comment]
That's great news! 
Is there somewhere where we can see this in action? It would be great to be able to select the density in the Angular Material website alongside the theme!

**Comment by critchie** [comment]
Several controls don't seem to honor the density settings such as Input. No matter what I try the input field does not shrink down like it does in the Material Design demo. And if I apply a density the floating label disappears.
https://m2.material.io/design/layout/applying-density.html#components

Did I miss something or is density not fully developed yet?

**Comment by JanneHarju** [comment]
@critchie Are you using Angular Material 15? If yes then I can tell what I noticed in todays researches. Class definition must be ether in global style files like this:
    .test-form-field {
      @include mat.form-field-density(-1);
    }
Or decorated with ::ng-deep if used in component. Otherwise it doesn't affect.
With value -1 it shrink 4px but as you said if using -2 value label disapears. I think it is bug and someone need to report it.
And here is one extra tip I can provide:
If you wan't use button density you need to set class definition to upper level element not to button itself.

    <div class="shrink-button">
        <button mat-raised-button >Push the button</button>
    </div>

and then style can be

    .shrink-button {
      @include mat.button-density(-1);
    }

So I think it is some kind of css selector problem in many cases.

**Comment by cluen** [comment]
Maybe @jelbourn can shed some light on this. 

Our complete application layout is broken after upgrading angular and material from 13 to 15. All components are larger, missing old defaults (i.e. appearance: fill | outline |none???). Trying to fix it with density ends in no labels for inputs. 
It's very useless to have 5 inputs among each other, without knownig what to type in?!

**Comment by seyfert** [comment]
@cluen You can read [the migration guide](https://material.angular.io/guide/mdc-migration) for breaking changes with mat-form-field.

The missing labels are due to [a setting in the MDC components](https://github.com/material-components/material-components-web/blob/master/packages/mdc-textfield/_variables.scss#L116) which Angular Material now uses.

If you add this to your global SCSS you can override the setting:
```
@use '@material/textfield' as mdc-textfield;

mdc-textfield.$minimum-height-for-filled-label: 40px;
```

This may not be officially supported so be aware that this could break in the future.

To get the "none" styling, for form fields inputs, you should [read the MDC documentation for styling a text field](https://github.com/material-components/material-components-web/blob/master/packages/mdc-textfield/README.md). You can create your own class to apply to a mat-form-field to format the inputs however you prefer using the defined mixins.
```
.my-custom-form-field {
    .mdc-text-field--filled {
        @include mdc-textfield.fill-color(transparent);
        @include mdc-textfield.disabled-fill-color(transparent);
        ....
    }
}
```

**Comment by amysorto** [comment]
I'm closing this issue since MDC components have been available since v15.  For any specific density issues, please file new issues for them, thanks so much!
